### PR TITLE
lightning: fix the bug that the GBK-encoded separator could not be parsed normally

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

The charset setting should be performed on all CSV parsers we create. Otherwise, when we try to check a GBK-encoded CSV file whose separator is `，` in GBK, lightning will report syntax error like:

```shell
syntax error: cannot have consecutive fields without separator
```

The data in UTF-8 looks like this:

image

We use `iconv` to transfer it into the GBK encoding.

```shell
iconv -c -f utf-8 -t gb18030 ../origin.csv > gbk_sbtest.gbk_sbtest1.000000000.csv
```

Try to import it whit the master version:

```shell
Verbose debug logs will be written to tidb-lightning.log

Error: restore table `gbk_sbtest`.`gbk_sbtest1` failed: in file gbk_sbtest.gbk_sbtest1.000000000.csv:0 at offset 6: syntax error: cannot have consecutive fields without separator
tidb lightning encountered error:  restore table `gbk_sbtest`.`gbk_sbtest1` failed: in file gbk_sbtest.gbk_sbtest1.000000000.csv:0 at offset 6: syntax error: cannot have consecutive fields without separator
```

### What is changed and how it works?

Construct a new charset convertor for every CSV parser, so that every call to the CSV parser could handle GBK encoding properly.

image

image

### Check List 

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)

See the test result above.

### Release note 

```release-note
None.
```

### CODE REVIEW

**Comment by JmPotato** [comment]
@kennytm @glorv @Little-Wallace PTAL, thx!

File: br/pkg/lightning/mydump/region.go [hunk 1]
```diff
@@ -359,7 +359,7 @@ func SplitLargeFile(
 		if err != nil {
 			return 0, nil, nil, err
 		}
-		parser, err := NewCSVParser(&cfg.Mydumper.CSV, r, int64(cfg.Mydumper.ReadBlockSize), ioWorker, true, nil)
```
Comment (kennytm) [inline br/pkg/lightning/mydump/region.go:362]: wouldn't actually constructing a `NewCharsetConvertor` replacing the `nil` here also fix the problem? and the diff would be much smaller.

Comment (JmPotato) [reply br/pkg/lightning/mydump/region.go:362]: My concern is that this way every time you build a CSV parser you need to consider whether to pass a charset convertor and if not, the charset setting may not work globally, which is quite like an undefined behavior. Putting its construction inside the `NewCSVParser` seems safer.

Comment (kennytm) [reply br/pkg/lightning/mydump/region.go:362]: it should always be passed? the only cases where you provide `nil` unconditionally is in tests.

Comment (JmPotato) [reply br/pkg/lightning/mydump/region.go:362]: Yes, you're right. So is passing it rather than making it inside more reasonable🤔？

Comment (kennytm) [reply br/pkg/lightning/mydump/region.go:362]: i just find the diff to be bigger than necessary 😅

**Review by kennytm — COMMENTED** [review]
LGTM but could you include a simple test case.

**Comment by JmPotato** [comment]
> LGTM but could you include a simple test case.

Added. PTAL.

**Review by kennytm — APPROVED** [review]

**Pull request merged**